File: scripts/backfill-lapsed-at.py
#!/usr/bin/env python3
"""
Backfill lapsed_at aus INPADOC LEGSTAT XMLs.
Liest outer ZIPs → inner ZIPs → XMLs per Streaming (kein RAM-Problem).
Extrahiert PG25-Events und setzt lapsed_at = frühestes PG25-Datum.

Usage:
    python3 scripts/backfill-lapsed-at.py [--dry-run] [--dir=~/Downloads/INPADOC]
"""

import sys, os, re, zipfile, psycopg2, io
from pathlib import Path
from collections import defaultdict

DRY  = "--dry-run" in sys.argv
DIR  = next((a.split("=",1)[1] for a in sys.argv if a.startswith("--dir=")),
            str(Path.home() / "Downloads/INPADOC"))
DIR  = str(Path(DIR).expanduser())
DB   = os.environ.get("DATABASE_URL", "postgresql://postgres:postgres@localhost:5433/patent_pilot")

# ── Regex (kompiliert, deutlich schneller) ─────────────────────────────────
DOC_RX   = re.compile(r'<legal-status-document[^>]*>([\s\S]*?)</legal-status-document>')
PUB_RX   = re.compile(r'<publication-reference[^>]*>[\s\S]*?<doc-number>(\d+)</doc-number>[\s\S]*?<kind>([^<]+)</kind>[\s\S]*?</publication-reference>')
EVT_RX   = re.compile(r'<legal-event[^>]*>([\s\S]*?)</legal-event>')
DATE_RX  = re.compile(r'<event-date>(\d{8})</event-date>')
PG25_STR = '<event-code>PG25</event-code>'

def extract_lapse_events(xml: str, country: str) -> list[tuple[str, str]]:
    results = []
    for doc_m in DOC_RX.finditer(xml):
        doc = doc_m.group(1)
        pubs = PUB_RX.findall(doc)
        patent_number = None
        for num, kind in pubs:
            if kind.startswith('B'):
                patent_number = f"{country}{num}{kind}"; break
        if not patent_number and pubs:
            patent_number = f"{country}{pubs[0][0]}{pubs[0][1]}"
        if not patent_number:
            continue

        lap_dates = []
        for ev_m in EVT_RX.finditer(doc):
            ev = ev_m.group(1)
            if PG25_STR not in ev:
                continue
            dm = DATE_RX.search(ev)
            if dm:
                d = dm.group(1)
                lap_dates.append(f"{d[:4]}-{d[4:6]}-{d[6:8]}")

        if lap_dates:
            lap_dates.sort()
            results.append((patent_number, lap_dates[0]))
    return results

# ── DB ─────────────────────────────────────────────────────────────────────
conn = psycopg2.connect(DB)
conn.autocommit = False
cur  = conn.cursor()

def flush(entries: list[tuple[str, str]]):
    if not entries or DRY:
        return len(entries)
    from psycopg2.extras import execute_values
    execute_values(cur, """
        UPDATE patents SET
            lapsed_at  = data.lapsed_at::date,
            updated_at = now()
        FROM (VALUES %s) AS data(patent_number, lapsed_at)
        WHERE patents.patent_number = data.patent_number
          AND patents.status = 'lapsed'
    """, entries)
    conn.commit()
    return len(entries)

# ── Main ───────────────────────────────────────────────────────────────────
print(f"Backfill lapsed_at | DIR={DIR} | DRY={DRY}\n")

outer_zips = sorted(
    p for p in Path(DIR).glob("backfile-legstat_xml_*.zip")
)
print(f"{len(outer_zips)} outer ZIPs\n")

total_events = 0
total_updated = 0

for outer_path in outer_zips:
    print(f"\n→ {outer_path.name}")
    with zipfile.ZipFile(outer_path, 'r') as outer:
        inner_names = [n for n in outer.namelist()
                       if '/DOC/LEGSTAT-' in n and n.endswith('.zip')]

        for inner_name in inner_names:
            # Country aus Filename: LEGSTAT-202608-EP-AppDate...
            country_m = re.search(r'LEGSTAT-\d+-([A-Z]+)-', inner_name)
            country = country_m.group(1) if country_m else 'XX'

            # Nur EP/DE/US verarbeiten (Rest überspringen)
            if country not in ('EP', 'DE', 'US'):
                continue

            # Inner ZIP streamen (kein RAM-Load des ganzen outer ZIP)
            inner_data = outer.read(inner_name)
            with zipfile.ZipFile(io.BytesIO(inner_data), 'r') as inner:
                xml_names = [n for n in inner.namelist() if n.endswith('.xml')]
                for xml_name in xml_names:
                    xml = inner.read(xml_name).decode('utf-8', errors='replace')
                    entries = extract_lapse_events(xml, country)
                    if entries:
                        updated = flush(entries)
                        total_updated += updated
                        total_events  += len(entries)
                        print(f"\r  {total_events:,} events → {total_updated:,} updates", end='', flush=True)

print(f"\n\nFertig: {total_events:,} Lapse-Events, {total_updated:,} DB-Updates.")
cur.close()
conn.close()

File: scripts/backfill-lapsed-at.ts
/**
 * Backfill lapsed_at from INPADOC LEGSTAT XMLs.
 *
 * Liest alle outer ZIPs → inner ZIPs → XMLs.
 * Extrahiert PG25 (= "Lapsed in a contracting state") events.
 * Setzt lapsed_at = frühestes PG25 Datum pro Patent.
 *
 * Usage:
 *   node --experimental-strip-types scripts/backfill-lapsed-at.ts \
 *     [--dir=~/Downloads/INPADOC] [--dry-run]
 */

// @ts-nocheck
import { readdirSync } from "fs";
import { join } from "path";
import AdmZip from "adm-zip";
import postgres from "postgres";

const DIR    = process.argv.find(a => a.startsWith("--dir="))?.slice(6)
  ?? (process.env.HOME + "/Downloads/INPADOC");
const DRY    = process.argv.includes("--dry-run");
const DB_URL = process.env.DATABASE_URL;
if (!DB_URL) { console.error("DATABASE_URL fehlt"); process.exit(1); }
const pg = postgres(DB_URL, { ssl: false, max: 5 });

// ── XML Parser ───────────────────────────────────────────────────────────────

interface LapsedEntry {
  patentNumber: string;  // e.g. EP0153093
  lapsedAt: string;      // YYYY-MM-DD (earliest PG25)
}

function extractLapseEvents(xml: string, country: string): LapsedEntry[] {
  const results: LapsedEntry[] = [];

  const docRx  = /<legal-status-document[^>]*>([\s\S]*?)<\/legal-status-document>/g;
  const pubRx  = /<publication-reference[^>]*>[\s\S]*?<doc-number>(\d+)<\/doc-number>[\s\S]*?<kind>([^<]+)<\/kind>[\s\S]*?<\/publication-reference>/g;
  const evtRx  = /<legal-event[^>]*>([\s\S]*?)<\/legal-event>/g;
  const dateRx = /<event-date>(\d{8})<\/event-date>/;

  let docMatch: RegExpExecArray | null;
  while ((docMatch = docRx.exec(xml)) !== null) {
    const doc = docMatch[1]!;

    // Publication-Referenz extrahieren (B-Kind bevorzugen)
    const pubRefs: string[] = [];
    pubRx.lastIndex = 0;
    let pr: RegExpExecArray | null;
    while ((pr = pubRx.exec(doc)) !== null) {
      pubRefs.push(`${country}${pr[1]}${pr[2]}`);
    }
    const patentNumber = pubRefs.find(p => /[B][12]?$/.test(p)) ?? pubRefs[0];
    if (!patentNumber) continue;

    // PG25-Events suchen — INNERHALB jedes einzelnen <legal-event>-Blocks
    const lapDates: string[] = [];
    evtRx.lastIndex = 0;
    let ev: RegExpExecArray | null;
    while ((ev = evtRx.exec(doc)) !== null) {
      const evText = ev[1]!;
      if (!evText.includes("<event-code>PG25</event-code>")) continue;
      const dm = dateRx.exec(evText);
      if (dm) {
        const raw = dm[1]!;
        lapDates.push(`${raw.slice(0, 4)}-${raw.slice(4, 6)}-${raw.slice(6, 8)}`);
      }
    }
    if (!lapDates.length) continue;

    lapDates.sort();
    results.push({ patentNumber, lapsedAt: lapDates[0]! });
  }

  return results;
}

// ── DB flush ─────────────────────────────────────────────────────────────────

async function flush(entries: LapsedEntry[]): Promise<number> {
  if (!entries.length || DRY) return entries.length;
  const nums  = entries.map(e => e.patentNumber);
  const dates = entries.map(e => e.lapsedAt);
  await pg`
    UPDATE patents SET
      lapsed_at  = t.lapsed_at::date,
      updated_at = now()
    FROM unnest(${nums}::text[], ${dates}::text[]) AS t(patent_number, lapsed_at)
    WHERE patents.patent_number = t.patent_number
      AND patents.status = 'lapsed'
  `;
  return entries.length;
}

// ── Main ─────────────────────────────────────────────────────────────────────

async function main() {
  console.log(`Backfill lapsed_at — INPADOC PG25 Events`);
  console.log(`Dir: ${DIR} | DRY: ${DRY}\n`);

  const outerZips = readdirSync(DIR)
    .filter(f => f.startsWith("backfile-legstat") && f.endsWith(".zip"))
    .sort()
    .map(f => join(DIR, f));

  console.log(`${outerZips.length} outer ZIPs gefunden\n`);

  let totalEntries = 0;
  let totalUpdated = 0;

  for (const outerPath of outerZips) {
    console.log(`\n→ ${outerPath.split("/").pop()}`);
    const outer = new AdmZip(outerPath);
    const innerEntries = outer.getEntries().filter(e =>
      e.entryName.includes("/DOC/LEGSTAT-") && e.entryName.endsWith(".zip")
    );

    for (const innerEntry of innerEntries) {
      const innerBuf = innerEntry.getData();
      const inner = new AdmZip(innerBuf);
      const xmlEntries = inner.getEntries().filter(e => e.entryName.endsWith(".xml"));

      for (const xmlEntry of xmlEntries) {
        const xml = xmlEntry.getData().toString("utf-8");
        // Extract country from filename e.g. LEGSTAT-202608-EP-...
        const country = xmlEntry.entryName.match(/LEGSTAT-\d+-([A-Z]+)-/)?.[1] ?? "EP";
        const entries = extractLapseEvents(xml, country);

        if (entries.length > 0) {
          const updated = await flush(entries);
          totalUpdated += updated;
          totalEntries += entries.length;
          process.stdout.write(`\r  ${totalEntries.toLocaleString()} lapse events → ${totalUpdated.toLocaleString()} DB updates`);
        }
      }
    }
  }

  console.log(`\n\nFertig: ${totalEntries.toLocaleString()} Lapse-Events, ${totalUpdated.toLocaleString()} Updates.`);
  await pg.end();
}

main().catch(e => { console.error(e); process.exit(1); });
